Add `where` predicate and Machine ID support to SSH host certificates

timothyb89 opened the pull request:
This adds `where` RBAC predicate support to `host_cert` pseudo-resources, and adds support for requesting host certs to Machine ID (with or without these RBAC predicates).

More specifically, this:
 * Makes 3 new set functions available during `where` predicate evaluation: `all_equal`, `all_end_with`, and `is_subset`. These make it possible to examine the `host_cert.principals` field.
 * Adds a new optional `host_cert` field to the standard `where` context. It's only ever set specifically in `GenerateHostCert()`.
 * Sets a custom context in `GenerateHostCert()` to pass along host cert parameters.
 * Makes current config available to tbot config templates
 * Adds a new `ssh_host_cert` config template to Machine ID
 * Refactors `tbot/config` tests to use shared mock bot / auth implementations.

Host certs can be requested in Machine ID by adding a new config template to `tbot.yaml`:
```yaml
destinations:
  - directory: /opt/machine-id
    configs:
      - ssh_host_cert:
          principals: [foo.example.com]
```

Bots will need a role like the following:
```yaml
kind: role
metadata:
  name: hostcert-bot
version: v5
spec:
  allow:
    rules:
      - resources:
          - host_cert
        verbs:
          - create
        where: 'is_subset(host_cert.principals, "foo.example.com")'
```

See the RFD for more information:
https://github.com/gravitational/teleport/blob/master/rfd/0083-machine-id-host-certs.md

timothyb89 committed:
Add `where` predicate and Machine ID support to SSH host certificates

This adds `where` RBAC predicate support to `host_cert`
pseudo-resources, and adds support for requesting host certs to
Machine ID (with or without these RBAC predicates).

More specifically, this:
 * Makes 3 new set functions available during `where` predicate
   evaluation: `all_equal`, `all_end_with`, and `is_subset`. These
   make it possible to examine the `host_cert.principals` field.
 * Adds a new optional `host_cert` field to the standard `where`
   context. It's only ever set specifically in `GenerateHostCert()`.
 * Sets a custom context in `GenerateHostCert()` to pass along host
   cert parameters.
 * Makes current config available to tbot config templates
 * Adds a new `ssh_host_cert` config template to Machine ID
 * Refactors `tbot/config` tests to use shared mock bot / auth
   implementations.

Host certs can be requested in Machine ID by adding a new config
template:
```yaml
destinations:
  - directory: /opt/machine-id
    configs:
      - ssh_host_cert:
          principals: [foo.example.com]
```

Bots will need a role like the following:
```yaml
kind: role
metadata:
  name: hostcert-bot
version: v5
spec:
  allow:
    rules:
      - resources:
          - host_cert
        verbs:
          - create
        where: 'is_subset(host_cert.principals, "foo.example.com")'
```

See the RFD for more information:
https://github.com/gravitational/teleport/blob/master/rfd/0083-machine-id-host-certs.md

timothyb89 committed:
Fix lint failure

timothyb89 committed:
Improve doc comments on the new predicate functions.

strideynet commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -3645,3 +3645,139 @@ func TestListResources_WithRoles(t *testing.T) {
+		where      string
+		deny       bool
+		denyWhere  string
+		expect     func(error) bool
```
nit: I'm not sure if anyone else gets this, but I do find the test table quite hard to read when the expect is omitted for success cases. Not a major issue, just something that came to mind.

timothyb89 replied:
I'll add a quick `noError` helper function.

strideynet replied:
Perfect thanks.

strideynet commented on lib/services/parser.go:
```diff
@@ -53,6 +53,93 @@ var (
+}
+
+// PredicateIsSubset determines if the first parameter is contained within the
+// variadic args. Only strings are supported; the first parameter may either
```
This comment could be a bit clearer, it says "only strings are supported" followed by "may either be a string or list of strings" - Maybe include something that suggests that the variadic args may only be strings.

timothyb89 replied:
That's fair, I'll update the comment.

strideynet commented on lib/services/parser.go:
```diff
@@ -53,6 +53,93 @@ var (
+// PredicateAllEqual is a custom function to test if all entries in a []string
+// are equal to a certain value. This is primarily useful for comparing string
+// fields that are only expected to contain a single, specific value.
+func PredicateAllEqual(a interface{}, b interface{}) predicate.BoolPredicate {
```
I'm split on whether or not this makes more sense as a seperate predicate, or if we should just modify `equals` to support this. I think you've made the right call, just figured its worth discussing.

timothyb89 replied:
I think that could make it hard (maybe impossible) to compare a plain `string` to a value that you want to contain a `string` but actually contains a `[]string`. I don't think this situation is ever likely to happen since we very rarely compare dynamic values, but it's not impossible.

timothyb89 committed:
Address review feedback

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/machine-id-host-certs

smallinsky commented on lib/tbot/config/configtemplate_ssh_host_cert.go:
```diff
@@ -0,0 +1,130 @@
+
+	// We'll need a client for the impersonated identity to request the certs.
+	botCfg := bot.Config()
+	authClient, err := bot.AuthenticatedUserClientFromIdentity(ctx, currentIdentity, botCfg.AuthServer)
```
Does the bot.Config needs to be called to get the `botCfg.AuthServer` in order to inject it to the `AuthenticatedUserClientFromIdentity` ? 
Why not just reference to bot.config.AuthServer inside the `AuthenticatedUserClientFromIdentity` function ?

timothyb89 replied:
Good point, we only ever pass it the configured value anyway. Unfortunately we still need to plumb it through to access `CertificateTTL`, though.

smallinsky commented on lib/tbot/config/configtemplate_ssh_host_cert.go:
```diff
@@ -0,0 +1,130 @@
+	defaultSSHHostCertPrefix = "ssh_host"
+)
+
+// TemplateSSHClient contains parameters for the ssh_config config
```
```suggestion
// TemplateSSHHostCert contains parameters for the ssh_config config
```

smallinsky commented on lib/tbot/config/configtemplate_ssh_host_cert.go:
```diff
@@ -0,0 +1,130 @@
+	return TemplateSSHHostCertName
+}
+
+func (c *TemplateSSHHostCert) Describe(destination bot.Destination) []FileDescription {
```
nit:  missing go doc.

smallinsky commented on lib/tbot/config/configtemplate_ssh_host_cert.go:
```diff
@@ -0,0 +1,130 @@
+	return ret
+}
+
+func (c *TemplateSSHHostCert) Render(ctx context.Context, bot Bot, currentIdentity *identity.Identity, destination *DestinationConfig) error {
```
nit: missing go docs.

smallinsky commented on lib/tbot/config/bot.go:
```diff
@@ -42,4 +48,145 @@ type Bot interface {
 	// Client retrieves the current auth client.
 	Client() auth.ClientI
+
+	// AuthenticatedUserClientFromIdentity returns a client backed by a specific
+	// identity.
+	AuthenticatedUserClientFromIdentity(ctx context.Context, id *identity.Identity, authServer string) (auth.ClientI, error)
```
Does AuthenticatedUserClientFromIdentity interface function is really needed. Right no it only referring to bot.Log to internal Bot filed to create the auth Client ?

timothyb89 replied:
Fair point, I'll move it into the `identity` package.

timothyb89 replied:
Hrm, on second thought, we need to be able to mock this in tests, so promoting it to a top-level function or otherwise splitting it up while keeping the function mockable will require more of a redesign than I think is worth pursuing at the moment.

timothyb89 commented on lib/tbot/config/bot.go:
```diff
@@ -42,4 +48,145 @@ type Bot interface {
+	Config() *BotConfig
+}
+
+const (
+	// mockProxyAddr is the address of the mock proxy server, used in tests
+	mockProxyAddr = "tele.blackmesa.gov:443"
+
+	// mockClusterName is the cluster name for the mock auth client, used in
+	// tests
+	mockClusterName = "tele.blackmesa.gov"
+)
+
+// mockAuth is a minimal fake auth client, used in tests
+type mockAuth struct {
+	auth.ClientI
+
+	clusterName string
+	proxyAddr   string
+	t           *testing.T
+}
```
Yeah, that was a dumb one on my part, will fix.

timothyb89 committed:
Address review feedback

- Remove `AuthenticatedUserClientFromIdentity`'s unnecessary
  AuthServer param
- Add missing doc comments
- Pull out `getTestIdent` from the Kubernetes template test
- Move various test helpers into their own file
- Unexport new predicate functions

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/machine-id-host-certs

timothyb89 committed:
Merge branch 'master' into timothyb89/machine-id-host-certs
